src/packages/connections/sockets/logic/creators/createSocketCommunicator.ts

```diff
@@ -113,6 +120,51 @@ export default function createSocketCommunicator({
     return send(data, processes, abort);
   }
 
+  async function scatter(
+    data: unknown[],
+    sendCount: number,
+    receiveBuf: unknown[],
+    receiveCount: number,
+    abort?: Abort,
+  ) {
+    const allDevicesLength = selfConnections.length + 1; // include myself
+    // validate input
+    if (sendCount * allDevicesLength > data.length) {
+      throw new Error('wrong sendCount');
+    }
+    if (receiveCount > sendCount) {
+      throw new Error('Cant receive more than send');
+    }
+
+    // split data to send
+    const splittedData: Array<Array<unknown>> = [];
+    for (let i = 0; i < allDevicesLength; i++) {
+      const start = i * sendCount;
+      const end = start + sendCount;
+      splittedData.push(data.slice(start, end));
+    }
+
+    const connectedProcesses = selfConnections.map(
+      ({ device: { process } }) => process,
+    );
+    const processes = [selfProcess, ...connectedProcesses];
+    await Promise.all(
+      processes.map((process, index) =>
+        send(splittedData[index], [process], abort),
+      ),
+    );
+
+    const { data: received, process } = await receive(abort);
+    if (!Array.isArray(received)) {
+      throw new Error(
+        `Received data must be an array. Got ${received} from ${process.code}`,
+      );
+    }
+    receiveBuf.push(...received.slice(0, receiveCount));
+
+    return
+  }
+
   function receive(abort?: Abort) {
```

Comment:
question(blocking): 

there is a conceptual trouble in the current logic. in this method, we just use simple assignments to get value, for example
```ts
const { data } = await communicator.receive();
```
. it means, that we can use `await` only when we can assign value somewhere. 

but in the `scatter` method we can specify the receiving buffer in the method call. so i think we should refactor the `receive` method in the same way.

what do you think?

Reply:
I absolutely agree, that is the problem I faced with buffer may be fixed with changing receive method